### PR TITLE
[OpenAI.Inference] Adding typespec definitions for DALL-E

### LINKED ISSUE #24499: [Azure OpenAI Service - Azure OpenAI Service] API Review
New API Review meeting has been requested.

**Service Name**: Azure OpenAI Service - Azure OpenAI Service
**Review Created By**: Gerardo Lecaros
**Review Date**: 07/25/2023 01:00 PM PT
**Onboarding Record**: https://dev.azure.com/azure-sdk/Release/_workitems/edit/9320
**PR**: https://github.com/Azure/azure-rest-api-specs/pull/24382
**Hero Scenarios Link**: Not Provided
**Core Concepts Doc Link**: Not Provided

**Description**: The objective of this meeting is to review the TypeSpec definitions that our team put together to match the service's defintitions for Azure DALL-E.

Detailed meeting information and documents provided can be accessed [here](https://web.powerapps.com/apps/e41c69de-ef96-44ba-b287-0e012e7c13ae?view=edit&event=e988c892-ca0e-ee11-8f6d-000d3a341bd5)
For more information that will help prepare you for this review, the requirements, and office hours, visit the documentation [here](https://aka.ms/azsdk/onboarding/restapischedule)

### PR DESCRIPTION
**Pull request by glecaros**
fix https://github.com/Azure/azure-rest-api-specs/issues/24499

### CODE REVIEW

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (trrwilson) [inline specification/cognitiveservices/OpenAI.Inference/models/images.tsp:16]: This could use some extra simplification/wordsmithing for user-facing docs clarity. "Represents the request data used to generate images" or similar?

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (trrwilson) [inline specification/cognitiveservices/OpenAI.Inference/models/images.tsp:18]: nitpick: "text" likely unnecessary ("A description" seems clear)

We just straight-up use "images" elsewhere and it's likely fine to snap to that vs. the more precise "image(s)"

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (trrwilson) [inline specification/cognitiveservices/OpenAI.Inference/models/images.tsp:22]: optional, defaults to 1

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (trrwilson) [inline specification/cognitiveservices/OpenAI.Inference/models/images.tsp:25]: optional, defaults to 1024x1024

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (trrwilson) [inline specification/cognitiveservices/OpenAI.Inference/models/images.tsp:28]: very optional

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (trrwilson) [inline specification/cognitiveservices/OpenAI.Inference/models/images.tsp:37]: I believe optional (omitted if successful, right?)

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (trrwilson) [inline specification/cognitiveservices/OpenAI.Inference/models/images.tsp:106]: Does this doc end up in emitted ref docs or are we purely limited to custom code for that? This doesn't seem user-facing as-is but that may be expected.

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:16]: updated

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:18]: updated both =)

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:22]: made optional and added the default to the doc

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:25]: made optional and added the default to the doc

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:28]: fixed

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:37]: fixed

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:106]: from what I could see, this type was not generated as is, so the docs wouldn't show up to the developer customer

File: specification/cognitiveservices/OpenAI.Inference/main.tsp
Comment (jpalvarezl) [inline specification/cognitiveservices/OpenAI.Inference/main.tsp:43]: shouldn't we be adding values to the versions enum instead of modifying? Also, are the other endpoints absolutely the same for `v2022_12_01` and `v2023_03_15_Preview` ?

Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/main.tsp:43]: oops, should have been 2023-05-15 (no preview)

Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/main.tsp:43]: the changes are just additive, we may want to keep all the non-preview versions, but preview versions sometime disappear.

File: specification/cognitiveservices/OpenAI.Inference/models/chat.completions.tsp
Comment (mikeharder) [inline specification/cognitiveservices/OpenAI.Inference/models/chat.completions.tsp:160]:
```suggestion
    #suppress "@azure-tools/typespec-autorest/union-unsupported" "https://github.com/Azure/typespec-azure/issues/3180"
```

File: specification/cognitiveservices/OpenAI.Inference/models/completions.create.tsp
Comment (mikeharder) [inline specification/cognitiveservices/OpenAI.Inference/models/completions.create.tsp:190]:
```suggestion
    #suppress "@azure-tools/typespec-autorest/union-unsupported" "https://github.com/Azure/typespec-azure/issues/3180"
```

File: specification/cognitiveservices/OpenAI.Inference/models/completions.create.tsp
Comment (mikeharder) [inline specification/cognitiveservices/OpenAI.Inference/models/completions.create.tsp:197]:
```suggestion
    #suppress "@azure-tools/typespec-autorest/union-unsupported" "https://github.com/Azure/typespec-azure/issues/3180"
```

**Comment by mikeharder** [comment]
@markcowl: Can you please take a look at the readme and breaking change errors in this PR?

**Comment by mikeharder** [comment]
@glecaros: Note that we now require all `*.tsp` files to be formatted correctly.  If you make any further changes to `*.tsp` files, please run `npx tsp format **/*.tsp` to ensure they are formatted correctly before pushing to your PR.

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (johanste) [inline specification/cognitiveservices/OpenAI.Inference/models/images.tsp:18]: I'd prefer to be explicit about the sizes because I somehow suspect that Large will not stay the biggest value forever.

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (johanste) [inline specification/cognitiveservices/OpenAI.Inference/models/images.tsp:29]: The comment does not match the enum value names. Which can be solved by not renaming the values in the enum.

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (johanste) [inline specification/cognitiveservices/OpenAI.Inference/models/images.tsp:37]: This model looks incomplete. What happens if you specify `response_format = "b64_json"`? Or have we explicitly decided not to support that?

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (johanste) [inline specification/cognitiveservices/OpenAI.Inference/models/images.tsp:66]: This is an optional property, no?

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:18]: yeah, couldn't think of a name that would actually work with the generators though (that's why I changed it)... when I tried `256x256`  for enum name C# generated something like `ImageSize._256x256` for the value to be a valid ID

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:37]: response_format is not part of the service API definition we based these on

Comment (johanste) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:37]: I don't know when it was added, but it is at least supported in OAI...

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:66]: fixed

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp]: updated

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (johanste) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:18]: I believe using a prefix like `Size` (or even repeating the "Image" part would be preferable. E.g. `ImageSize.Size256x256` or `ImageSize.ImageSize256x256`.

Comment (trrwilson) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:18]: @johanste do you have a recommendation on a way we could keep this pretty across languages? I agree with you on the fragility of the generic labels; I'd love if we could just encode this as proper width and height parameters that we combine ourselves if needed, but that comes with its own drawbacks with how easily invalid input becomes to provide.

We could have a middle ground of something like "Size256x256" -- that just feels a little bit ugly.

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:37]: yeah, for AOAI it is not there in the latest service definitions, we can add it now or when we do the work to support the vanilla (non-LRO) generation API

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (johanste) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:18]: Across all languages? Not sure - but if I was expressing this in a TypeScript library, I'd probably prefer to use a union and completely ignore the labels. Same for Python. And thus avoiding the name/enumness altogether.

Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp]: updated with that

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (johanste) [inline specification/cognitiveservices/OpenAI.Inference/models/images.tsp:27]: We should provide the default value if there is one.

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:29]: changed the enum values back to the sizes

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (johanste) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:37]: Since we know it is coming, we should at least spend a couple of minutes to make sure we are not digging ourselves into a hole w. regards to structure or naming. E.g. maybe we want to have an ImageLocation rather than Image name for the returned model if we expect to provide differently named methods for the b64 version and the url version.

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (johanste) [inline specification/cognitiveservices/OpenAI.Inference/models/images.tsp:22]: Do we really need this model? In other words, why isn't it an alias?

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:37]: Looking at the OAI swagger, it seems that the change to support the b64 format would be compatible with the current structure. When that comes, I imagine the model would be close to this:

```
model ImageResult {
  url?: url,
  b64_json?: string
}
```

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:22]: ImageRequest? What would you have it be an alias of?

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (johanste) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:37]: The proposal above misses two important pieces of information; they are mutually exclusive, and exactly one of the properties must be provided. A more precise description (names subject to change) is:

```
model ImageLocation {
    url: url;
}

model ImageData {
    b64_json: string;
}

alias ImageResult = ImageData | ImageLocation;
```

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (johanste) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:22]: An anonymous model. Or provide the model inline in the template operation.

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:27]: is there an annotation to do that? in this case the default is implicit, if not present, the service will use `1`

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:22]: well, image request we need, we could remove the ImageResponse model (we are not using it actually as we couldn't use two different models for the start and polling operations)

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:37]: that is fair, that'd be an improvement over how OpenAI defined their API. I'll for now rename ImageResult to ImageLocation. When/if we use the union we may need to figure out how to make that work with the openapi generation requirement the repo has.

I'll also ask to the service if they plan to ever support base64 output for the LRO operation.

Comment (johanste) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp]: Yeah, openapi2 didn't have a good way to express this. In OpenAPI 3+, it would be a `oneOf`. The closest thing we have in autorest/azure swagger land is to make it an "overload" (using the x-ms-paths vendor extension in our swagger) and have two different response models. Which is a pain (to say the least).

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:22]: well, this is GenerateImage's version of `ChatCompletionsOptions`, maybe we should rename it to `GenerateImageOptions`?

Comment (johanste) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp]: I would have the same feedback for `ChatCompletionsOptions`. As a service concept, these are just the parameters to the chat completions endpoint, they don't really "deserve" their own conceptual name - the bag of properties really only exist to be passed in as parameters to the method.

Comment (trrwilson) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp]: @johanste is there a way we can flatten this where appropriate and bundle where it's not?

I think "how many parameters we put into a function/method" has a lot of language specificity; e.g. in Python with its strong named parameter integration, I think we like "flat, even if numerous": https://azure.github.io/azure-sdk/python_design.html#service-method-parameters -- while in e.g. Java, we want to limit the number of parameters for "complex" methods by representing the request payload in a Options type: https://azure.github.io/azure-sdk/java_introduction.html#service-method-parameters

All of these OpenAI routes either do or appear that they easily could grow into the "complex" territory (6+ configurable things as per those Java guidelines), so having a way to flexibly represent things "flat" where they be flat and "bundled" where they should be bundled would be a big help.

Comment (johanste) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp]: This is where additional client codegen specific metadata comes in since that is client codegen (and possibly even language specific client code gen) choices to be made - as you call out. At the conceptual level for the service, this "thing" doesn't really have a name/identity other than it being the set of parameters to the endpoint. And as such, we prefer to keep it unnamed, which allows emitters to give it a name or not. This is doubly true for input parameters where emitters can almost always expand (through overloads or similar) without introducing breaking changes.

Comment (trrwilson) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp]: That makes sense. Is there a way we can accomplish that directly here? Specifically: if we get rid of the named model in the typespec and inline it, can we make Java and .NET still wrap it appropriately?

If we can snap to it now, that's certainly ideal. If it needs extra legwork, I'd like to move forward and fix/flatten this in one swoop (including for e.g. ChatCompletions as you mention).

Comment (johanste) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp]: My first question is if the clients actually want to wrap it/the whole thing in the first place. My suspicion is that at least the C# architect(s) might - in an ideal world - prefer it to be split and have the required input (e.g. `prompt`/`input` and `model` / where applicable) be positional parameters and have a separate split/partial model for the other more esoteric parameters. 

@KrzysztofCwalina, I think this could be a good discussion for you to pay attention to as well because it does play into what code to generate under what circumstances for C# (and if it should be called something something `Options` :)).

Comment (trrwilson) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp]: We do have a degree of that today with `deploymentOrModelName` being a required parameter on `GetCompletions`/`GetChatCompletions`/etc. in C#, with `prompt` still relegated to the complex `Options` by virtue of not being *strictly* required for pre-chat Completions.

For the purposes of moving forward with as-is DALL-E support, would it be appropriate for us to snap to that pattern of "required settings promoted to parameters, optional ones relegated to options" for `GenerateImages`? That establishes an even more anemic Options container, but (a) it's consistent and (b) there's a reasonable supposition that the collection of optional parameters will grow over time.

As much as we can, I'd like to defer the very good and very important discussion of overall representation to a broader audience and forum than the DALL-E feature--both to ensure we get the right voices heard for the bigger topic and to unblock this parity feature having client library representation. @johanste, what do you see as must-have to unblock this PR merging here?

Thanks.

File: specification/cognitiveservices/OpenAI.Inference/routes.tsp
Comment (markcowl) [inline specification/cognitiveservices/OpenAI.Inference/routes.tsp:20]: This `@added` is not needed - every type without a versioning decorator is included in the first api-version

File: specification/cognitiveservices/OpenAI.Inference/routes.tsp
Comment (markcowl) [inline specification/cognitiveservices/OpenAI.Inference/routes.tsp:34]: Same comment

Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/routes.tsp]: fixed

**Comment by mikeharder** [comment]
@glecaros: I updated the triggers for our CI pipeline to automatically run against PRs to your spec folder.  Please ensure this check is green before merging.

image

File: specification/cognitiveservices/OpenAI.Inference/routes.tsp
Comment (johanste) [inline specification/cognitiveservices/OpenAI.Inference/routes.tsp:55]:
```suggestion
#suppress "@azure-tools/typespec-azure-core/no-rpc-path-params" "Allowed because this is a non-standard status polling operation."
```

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (trrwilson) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:27]: Paraphrasing: it's a service-inferred default rather than a client-coded one. The longer version of this might be something like "If not provided, the image generation request operation will behave as if n=1 and generate a single image."

**Comment by glecaros** [comment]
> @glecaros: I updated the triggers for our CI pipeline to automatically run against PRs to your spec folder. Please ensure this check is green before merging.
> 
> image

thanks @mikeharder! Now the only checks (aside from the breaking change check), now the only ones left are: `azure-sdk-for-js`, `azure-sdk-for-net-track2`, and `azure-sdk-for-java`, do you know how can I fix those?

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (johanste) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:27]: n?: int32 = 1;

File: specification/cognitiveservices/OpenAI.Inference/main.tsp
Comment (trrwilson) [inline specification/cognitiveservices/OpenAI.Inference/main.tsp:54]: Note: open followup issue on the error generated when using `@visibility(read)`

**Comment by mikeharder** [comment]
> > @glecaros: I updated the triggers for our CI pipeline to automatically run against PRs to your spec folder. Please ensure this check is green before merging.
> > image
> 
> thanks @mikeharder! Now the only checks (aside from the breaking change check), now the only ones left are: `azure-sdk-for-js`, `azure-sdk-for-net-track2`, and `azure-sdk-for-java`, do you know how can I fix those?

The check for `azure-sdk-for-net-track2` actually passed, but status never made it from DevOps to GitHub.  This happens sometimes and unfortunately there's not much we can do about it.

The check for `azure-sdk-for-java` appears to have hung trying to clone a repo from GitHub, I will re-run it.

The check for `azure-sdk-for-js` did fail, I will investigate this one, but I'm not sure if all of the `azure-sdk-for-*` checks are even working right now.

**Comment by mikeharder** [comment]
> The check for `azure-sdk-for-net-track2` actually passed, but status never made it from DevOps to GitHub. This happens sometimes and unfortunately there's not much we can do about it.
> 
> The check for `azure-sdk-for-java` appears to have hung trying to clone a repo from GitHub, I will re-run it.
> 
> The check for `azure-sdk-for-js` did fail, I will investigate this one, but I'm not sure if all of the `azure-sdk-for-*` checks are even working right now.

The check for `azure-sdk-for-net-track2` failed now.  Not sure how it passed earlier but it doesn't matter.  Current failure is caused by this compiler error:

```
/mnt/vss/_work/1/s/azure-sdk-for-net/sdk/openai/Azure.AI.OpenAI/src/Custom/StreamingChoice.cs(37,65):
error CS1061: 'CompletionsFinishReason' does not contain a definition for 'Value' and no accessible extension
method 'Value' accepting a first argument of type 'CompletionsFinishReason' could be found (are you
missing a using directive or an assembly reference?)
[/mnt/vss/_work/1/s/azure-sdk-for-net/sdk/openai/
Azure.AI.OpenAI/src/Azure.AI.OpenAI.csproj::TargetFramework=netstandard2.0]
```

**Comment by mikeharder** [comment]
Check `azure-sdk-for-java` is failing with this error:

```
10:57:42.997 cmdout 	[generate.py] [ERROR] /mnt/vss/_work/1/s/azure-sdk-for-java/sdk/openai/
azure-ai-openai/src/main/java/module-info.java:[7,39] module not found: com.azure.core.experimental
```

@srnagar: Are you the right person to assist with this error?

**Comment by mikeharder** [comment]
Check `azure-sdk-for-js` failed with this error:

```
npx tsp compile /mnt/vss/_work/1/s/azure-sdk-for-js/sdk/openai/azure-ai-openai/TempTypeSpecFiles/OpenAI.Inference/main.tsp --emit @azure-tools/typespec-ts --option @azure-tools/typespec-ts.emitter-output-dir=/mnt/vss/_work/1/s/azure-sdk-for-js/sdk/openai/azure-ai-openai/
TypeSpec compiler v0.45.0
Failed to format file: /mnt/vss/_work/1/s/azure-sdk-for-js/sdk/openai/azure-ai-openai/src/rest/models.ts
ExternalError: Emitter "@azure-tools/typespec-ts" failed!
File issue at https://github.com/Azure/autorest.typescript/issues
Error: Cannot find module './parser-typescript.js'
```

@deyaaeldeen: Are you the right person to assist with this?

**Comment by mikeharder** [comment]
@raych1: Three of the "sdk" checks are failing for this PR.  Are these checks expected to be passing now on valid PRs, so if these checks are failing it means something is wrong in the PR?  Or are these checks still not working reliably?

**Review by markcowl — APPROVED** [review]

**Comment by raych1** [comment]
> @raych1: Three of the "sdk" checks are failing for this PR. Are these checks expected to be passing now on valid PRs, so if these checks are failing it means something is wrong in the PR? Or are these checks still not working reliably?

@mikeharder , all the emitter are commented out except autorest. That's why SDK CI checks are skipped.

File: specification/cognitiveservices/OpenAI.Inference/routes.tsp
Comment (weidongxu-microsoft) [inline specification/cognitiveservices/OpenAI.Inference/routes.tsp:80]: Do we have a runtime log on the LRO responses? Or at least give us a response of last LRO poll.

Ask because the LRO may not be correctly defined.

It may need to be written as
```
op startGenerateImage is OaiLongRunningRpcOperation<
  ImageGenerationOptions,
  ImageOperationResponse,
  ImageResponse
>;
```

The `TStatusResult` in template is the model in "result" property. It is a **result**, not a **status**.
https://github.com/Azure/typespec-azure/blob/main/packages/typespec-azure-core/lib/foundations.tsp#L37-L51

Comment (weidongxu-microsoft) [reply specification/cognitiveservices/OpenAI.Inference/routes.tsp:80]: Or you may want to use [Azure.Core.LongRunningRpcOperation](https://github.com/Azure/typespec-azure/blob/main/packages/typespec-azure-core/lib/operations.tsp#L37-L63)

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (archerzz) [inline specification/cognitiveservices/OpenAI.Inference/models/images.tsp:129]: I think this is a typo, we should use `TResponse` instead of `ImageOperationResponse`.
```suggestion
    TResponse &
```

**Comment by mikeharder** [comment]
> > @raych1: Three of the "sdk" checks are failing for this PR. Are these checks expected to be passing now on valid PRs, so if these checks are failing it means something is wrong in the PR? Or are these checks still not working reliably?
> 
> @mikeharder , all the emitter are commented out except autorest. That's why SDK CI checks are skipped.

The emitters were commented out **because** the CI checks were failing earlier.  We decided to merge this PR without the language emitters for now, and we will try to re-enable the language emitters in subsequent PRs, where we will need to investigate any failures.

File: specification/cognitiveservices/OpenAI.Inference/models/images.tsp
Comment (johanste) [inline specification/cognitiveservices/OpenAI.Inference/models/images.tsp:97]: This is problematic if it is a terminal state. Need to follow up with the service team what this state actually represents.

Comment (glecaros) [reply specification/cognitiveservices/OpenAI.Inference/models/images.tsp:97]: deleted was just there as because we were reusing then enum definition used in authoring. It was removed in a follow-up PR

**Pull request merged**
